Use remark for easily detectable markdown documentation

halkeye opened the pull request:
Original plan was to just support the wiki export, but was able to support a bunch more. I really just wanted to see if I could do it based on what I learned on the gatsby-jenkins.io project. Pull out some markdown reading/rendering from api to local. Did this by marking specific content as text/markdown and letting the gatsby-transform-remark handle converting it. There's tons of really cool remark plugins like the auto linking headers\.

Changes:

* Upgrade gatsby (cause didn't want to figure out what versions of remark worked with older version)
* Create Plugin Wiki objects that will be processed by markdown / remark  handler
* Link wiki objects to plugins
* Render wiki objects
* Syntax highlighting for code blocks
* Resize any youtube videos to fit the screen
* Autolink headers (ex http://next-cherries.surge.sh/scp/#version-19---)

Major wins.
* As there's no github API calls, its hitting the raw pages, it is much more reliable at getting the docs.
* Reduced plugin-site-api usage
* SO MANY GATSBY-REMARK PLUGINS - https://www.gatsbyjs.com/plugins/?=remark

Marking it as a draft as I kinda want to play around with it more, but open to feedback if its a good idea or not
cc @MarkEWaite @zbynek

halkeye committed:
Pull out some markdown reading/rendering from api to local

* Upgrade gatsby (cause didn't want to figure out what versions of
  remark worked with older version)
* Create Plugin Wiki objects that will be processed by markdown / remark
  handler
* Link wiki objects to plugins
* Render wiki objects

halkeye committed:
createNodeId produces the same id for the same input

halkeye committed:
Actually support any plugin that directly has README.md file as its url

halkeye commented:
![image](https://user-images.githubusercontent.com/110087/142829804-8d81960e-0b8d-4ba9-bbd9-1bc19397ee90.png)
^^^ plugin site for /scp at time of writing

halkeye committed:
Fix lint warning

timja commented on gatsby-config.js:
```diff
@@ -42,7 +42,48 @@ module.exports = {
+                    }
+                },
+                {
+                    resolve: 'gatsby-remark-autolink-headers',
```
❤️  excited for this

MarkEWaite commented:
Strong +1 from me.  @stackscribe asked in Docs office hours last week about linking to locations inside a markdown page from other pages (like www.jenkins.io).  This looks like it will provide that.  Thanks for doing it!

zbynek commented:
I'm sure there are a lot simpler solutions for "autolink headers" :) But this looks like a useful step towards simpler architecture. Any thoughts on AsciiDOC ([example](https://github.com/jenkinsci/git-plugin/blob/master/README.adoc))? Do you also want to parse it via some gatsby plugin?

halkeye commented:
yea, the autolink headers that work on jenkins.io should work on plugins if we included it, but it seems the github HTML injects random HTML in there too.

https://www.gatsbyjs.com/plugins/gatsby-transformer-asciidoc/ does work, not fast or well, but does work. I'm using it on the jenkins.io gatsby project

halkeye committed:
fix tests i think

halkeye committed:
Switch to a fallthrough method of pulling down readme data

halkeye committed:
use blobs not try to serve github.com

halkeye commented:
oh wow, I think this might be more important than I originally thought. https://plugins.jenkins.io/scp/ is failing 24 hours later. I'm guessing plugin-site-api just doesn't handle things.

halkeye commented:
I've updated https://next-cherries.surge.sh with my latest build, so feel free to poke around and see if anything looks obviously broken.

timja commented:
I can’t see any docs at least on my phone 

https://next-cherries.surge.sh/azure-credentials/

I tried slack plugin as well

halkeye commented:
Thx I'll take a look tomorrow. I may have built it wrong locally. Gotta get
my demo site back up.

On Mon., Nov. 22, 2021, 11:33 p.m. Tim Jacomb, ***@***.***>
wrote:

> I can’t see any docs at least on my phone
>
> https://next-cherries.surge.sh/azure-credentials/
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkins-infra/plugin-site/pull/845#issuecomment-976223923>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA24B6LKMK3ULVR7ASCNY3UNM7T5ANCNFSM5IQMTKUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

halkeye committed:
Fix displaying non markdown

* Handle both markdown by a creating a field extension that will return
  the first of whichever fields is defined (so we don't need to have
          markdown and internal content in frontend's json)

halkeye commented:
Yea totally found the issue, was only rendering markdown, not (markdown or regular HTML) (plugin site api)

timja commented:
Where's my heading links!
https://next-cherries.surge.sh/azure-credentials/

Content is there now

halkeye commented:
> Where's my heading links!

sadly that wasn't the goal of this PR, the goal was to move some of the processing out of plugin-site-api, especially now how often I'm seeing failed documentation.

Next steps would be to move more of it into the system. (Handle ones that are specifically .adoc, then handle ones that are just root github, then see whats left over)

I'll do a more permanent fix for headers after this.

I'm leaning towards merging to fix how often I'm seeing docs failing to load
